flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/RetryTestExecutionExtension.java

```diff
@@ -0,0 +1,121 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.testutils.junit;
+
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.ConditionEvaluationResult;
+import org.junit.jupiter.api.extension.ExecutionCondition;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.TestExecutionExceptionHandler;
+import org.opentest4j.TestAbortedException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+
+import static org.apache.flink.testutils.junit.RetryExtension.RETRY_KEY;
+import static org.apache.flink.testutils.junit.RetryExtension.RETRY_NAMESPACE;
+import static org.apache.flink.testutils.junit.RetryExtension.getTestMethodKey;
+
+/** Extension to decide whether a retry test should run. */
+public class RetryTestExecutionExtension
+        implements ExecutionCondition, TestExecutionExceptionHandler, AfterEachCallback {
+    public static final Logger LOG = LoggerFactory.getLogger(RetryTestExecutionExtension.class);
+    private final int retryIndex;
+    private final int totalTimes;
+    private Class<? extends Throwable> repeatableException;
+
+    public RetryTestExecutionExtension(
+            int retryIndex, int totalTimes, Class<? extends Throwable> repeatableException) {
+        this.retryIndex = retryIndex;
+        this.totalTimes = totalTimes;
+        this.repeatableException = repeatableException;
+    }
+
+    @Override
+    public ConditionEvaluationResult evaluateExecutionCondition(ExtensionContext context) {
+        Map<String, Boolean> shouldRetry =
+                (Map<String, Boolean>) context.getStore(RETRY_NAMESPACE).get(RETRY_KEY);
+        String method = getTestMethodKey(context);
+        if (!shouldRetry.get(method)) {
+            return ConditionEvaluationResult.disabled(method + "have already passed or failed.");
+        }
+        return ConditionEvaluationResult.enabled(
+                String.format("Test %s: [%d/%d]", method, retryIndex, totalTimes));
+    }
+
+    @Override
+    public void handleTestExecutionException(ExtensionContext context, Throwable throwable)
+            throws Throwable {
+        String method = getTestMethodKey(context);
+        if (repeatableException != null) {
```

Comment:
are there any limitation as to what we can add to the ExtensionContext store? It would be quite neat if we could just encapsulate the 2 restart mechanisms in 2 implementations of a sort of restart strategy, and we just store one such instance in the store.

Reply:
I'll try to refact it , it seems not to have this limit.